### PR TITLE
DB Optimizations

### PR DESCRIPTION
**Pull request by phlipper**
This branch contains some database improvements including eliminating n+1 queries and adding some cache_counter columns.

These changes eliminate 100's of queries per request for large installations. 

More work remains for future pull requests.

### CODE REVIEW

File: spec/models/post_spec.rb [hunk 1]
```diff
@@ -65,6 +65,59 @@
     end
   end
 
+  context "Forum counters" do
+    describe "#update_forum_posts_count" do
+      it "increments #posts_count on create" do
+        forum = FactoryGirl.create(:forum)
+        count = forum.posts_count
+        new_topic = FactoryGirl.create(:topic, forum: forum)
```
Comment (parndt) [inline spec/models/post_spec.rb]: Can you please use Ruby 1.8 hashes here? Thanks.

File: spec/models/post_spec.rb [hunk 2]
```diff
@@ -65,6 +65,59 @@
...
+        forum.reload.posts.count.should == count + 1
+        forum.reload.posts_count.should == count + 1
+      end
+
+      it "does not increment #posts_count on save" do
+        forum = FactoryGirl.create(:forum)
+        count = forum.posts_count
+        new_topic = FactoryGirl.create(:topic, forum: forum)
```
Comment (parndt) [inline spec/models/post_spec.rb]: Can you please use Ruby 1.8 hashes here? Thanks.

**Comment by phlipper** [comment]
@parndt I have updated the 1.8-style hash rockets. Slip of habit.

**Comment by phlipper** [comment]
@parndt I will look in to why some of the test runs pass on Travis and some do not.

**Comment by parndt** [comment]
Thank you :)

**Comment by parndt** [comment]
The specs seem happy in sqlite but no other DB.

**Comment by radar** [comment]
Patch looks great. Thank you :+1:!

Could you look into why the specs are failing on PostgreSQL and MySQL please? If you fix these the I would be happy to merge in these changes.

**Comment by knewter** [comment]
@phlipper Any ideas what the problem might be on the other databases?  I'd love to see this merged...

**Comment by parndt** [comment]
ac59874f696be83080df744a9f57dcb830dd2af2 broke compatibility

**Comment by parndt** [comment]
@phlipper any chance you could rebase this against the current master (some of the functionality has been refactored already) and fix the DB incompatibilities please?

Note to future maintainer/@parndt: close this if another few weeks goes by without it being rebased and having passing tests

**Comment by daniely** [comment]
i plan on tackling this over the weekend. if there's any reason i should not please let me know

**Pull request closed without merging**